x/observer/keeper/ballot.go

```diff
@@ -82,28 +82,58 @@ func (k Keeper) AddBallotToList(ctx sdk.Context, ballot types.Ballot) {
 	k.SetBallotList(ctx, &list)
 }
 
-// ClearMaturedBallotsAndBallotList deletes all matured ballots and the list of ballots for a given height.
-// It also emits an event for each ballot deleted.
-func (k Keeper) ClearMaturedBallotsAndBallotList(ctx sdk.Context, maturityBlocksParam int64) {
-	maturedBallotsHeight := getMaturedBallotHeight(ctx, maturityBlocksParam)
-
-	// Fetch all the matured ballots, return if no matured ballots are found
-	// For the current implementation, this should never happen as ClearMaturedBallotsAndBallotList is only called after the Distribution of the rewards,
-	//	which means that there are matured ballots to be deleted
+// ClearFinalizedMaturedBallots deletes all matured and finalized ballots for a given height.
+// It emits an event for each ballot deleted.
+// If all ballots are finalized and deleted then the ballot list is also deleted.
+func (k Keeper) ClearFinalizedMaturedBallots(ctx sdk.Context, maturityBlocks int64) {
+	maturedBallotsHeight := getMaturedBallotHeight(ctx, maturityBlocks)
 	maturedBallots, found := k.GetBallotListForHeight(ctx, maturedBallotsHeight)
 	if !found {
 		return
 	}
-	// Delete all the matured ballots and emit an event for each ballot deleted
+
+	ballotsDeleted := 0
+	for _, ballotIndex := range maturedBallots.BallotsIndexList {
+		ballot, foundBallot := k.GetBallot(ctx, ballotIndex)
+		if !foundBallot {
+			continue
+		}
+		if !ballot.IsFinalized() {
+			continue
+		}
+		k.DeleteBallot(ctx, ballotIndex)
+		logBallotDeletion(ctx, ballot)
+		ballotsDeleted++
+	}
+
+	// If all ballots are finalized and deleted, then delete the list of ballots.
+	// This would avoid a second iteration at the buffered maturity height
+	// If there are a few ballots remaining we would delete the list at the buffered maturity height
+	if ballotsDeleted == len(maturedBallots.BallotsIndexList) {
+		k.DeleteBallotListForHeight(ctx, maturedBallotsHeight)
+	}
```

Comment:
We can consider updating the ballotListforHeight instead to remove deleted ballots, but that would involve setting the ballotList back to the store. This option looks cleaner, in my opinion